fastreplacestring/fastreplacestring.cpp

```diff
@@ -0,0 +1,171 @@
+#include <cstdio>
+#include <cstring>
+#include <sys/stat.h>
+#include <stdlib.h>
+#include <vector>
+
+#include <caml/alloc.h>
+#include <caml/memory.h>
+#include <caml/mlvalues.h>
+
+using namespace std;
+
+#define REHASH(a, b, h) ((((h) - (a)*d) << 1) + (b))
+
+// Rabin-Karp search (modified from:
+// http://www-igm.univ-mlv.fr/~lecroq/string/node5.html#SECTION0050)
+int indexOf(const char *needle, size_t needleLen, const char *haystack,
+            size_t haystackLen) {
+  if (needleLen == 0) {
+    return 0;
+  } else if (needleLen > haystackLen) {
+    return -1;
+  }
+
+  int d, hx, hy, i, j;
+  /* Preprocessing */
+  /* computes d = 2^(m-1) with
+    the left-shift operator */
+  for (d = i = 1; i < needleLen; ++i)
+    d = (d << 1);
+
+  for (hy = hx = i = 0; i < needleLen; ++i) {
+    hx = ((hx << 1) + needle[i]);
+    hy = ((hy << 1) + haystack[i]);
+  }
+
+  /* Searching */
+  j = 0;
+  while (j <= haystackLen - needleLen) {
+    if (hx == hy && memcmp(needle, haystack + j, needleLen) == 0) {
+      return j;
+    }
+    hy = REHASH(haystack[j], haystack[j + needleLen], hy);
+    ++j;
+  }
+
+  return -1;
+}
+
+int replace(const char *filename, const char *old, const char *newWord) {
+  FILE *in = fopen(filename, "rb");
+
+  // Check if file exists and can is read-write
+  // This is actually a shortcut because fopen might fail for a number of
+  // reasons
+  if (in == NULL) {
+    /* return "error: file doesn't exist\n"; */
```

Comment:
There were a few TODOs I had here:
- Make sure we call `fclose` in all the places we return - otherwise, we could leave file descriptors open in the case of an error.
- If we're planning to continue to use the FFI, we might consider return a real error object out of `replace` (I commented out the error messages at the moment), but we can pass back a variant `ReplaceSuccess` or `ReplaceFailure(string)` from here.
- Still need to implement replacing files in a directory, vs a specific file.